contracts/contracts/EventManager.sol

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+contract EventManager {
+    struct Event {
+        uint256 eventId;
+        uint16 eventType;
+        bytes data;
+    }
+
+    mapping(uint => Event[]) private _eventsByType;
+    uint private _eventBufferSize;
+    address public writer;
+
+    constructor(uint bufferSize, address _writer) {
+        _eventBufferSize = bufferSize;
+        writer = _writer;
+    }
+
+    function insertEvent(uint256 eventId, uint16 eventType, bytes memory data) internal {
+        require(_eventsByType[eventType].length < _eventBufferSize, "Event buffer is full");
+        _eventsByType[eventType].push(Event(eventId, eventType, data));
+    }
+
+    function eraseEvents(uint256[] memory eventIds, uint16 eventType) external {
+        require(msg.sender == writer, "Only writer can erase events");
+
+        Event[] storage events = _eventsByType[eventType];
+        uint256 i = 0;
+
+        while (i < events.length) {
+            bool found = false;
+            for (uint256 j = 0; j < eventIds.length; j++) {
```

Comment:
iteration over events is unbounded now. iirc, we decided to add a max cap for the number of events to prevent DoS.

Reply:
but we have `require(_eventsByType[eventType].length < _eventBufferSize, "Event buffer is full");` assertion to prevent DoS attack